README.md
# PyAnchorKnit
Inspired by the work of [Petros Vrellis](https://www.instagram.com/pvrellis/). Based on [Weaver](https://github.com/alyyousuf7/Weaver).

## Installation
### Prerequisites
`pyanchorknit` uses `pycairo`, which requires `pkg-config` and `cairo`. Check [here](https://pycairo.readthedocs.io/en/latest/getting_started.html) for installation commands for your platform.

### Install `pyanchorknit`
Usage of [pipx](https://pypa.github.io/pipx/) is encouraged.

```bash
pipx install pyanchorknit
```

## Example usage
```bash
pyanchorknit imgs/Johannes-Vermeer-Girl-With-a-Pearl-Earring.jpg --n-jobs 16 --n-edges 512 --maxlines 2000 --img-out imgs/
```

### Output image
<img src="imgs/Johannes-Vermeer-Girl-With-a-Pearl-Earring-weave.png" alt="drawing" width="400"/>

### Output JSON
```json
{
  "points": {
    "0": [
      799,
      400
    ],
    "1": [
      682,
      682
    ],
    ...,
    "7": [
      682,
      117
    ]
  },
  "traces": [
    [
      [
        0,
        4
      ],
      145236
    ],
    [
      [
        4,
        1
      ],
      113064
    ],
    ...,
    [
      [
        2,
        6
      ],
      130755
    ]
  ]
}
```

Which indicates the position of anchor points and traces (along with their distance).

# Dev install
## Poetry
### Install poetry
```bash
pipx install poetry
```

### Install the project
```bash
poetry install
```

## Install [pre-commit](https://pre-commit.com) hooks
```bash
poetry run pre-commit install
poetry run pre-commit install -t pre-push
```

## Publish
```bash
poetry publish --build
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] - 2021-10-02
### Changed
- Draw final image with Cairo

## [0.2.0] - 2021-10-02
### Added
- intensity threshold added as parameter

## [0.1.1] - 2021-10-02
### Fixed
- Traces JSON key fixed

## [0.1.0] - 2021-10-02
### Added
- Initial weaving algorithm


tests/test_pyweaving.py
from pyweaving import __version__


def test_version():
    assert __version__ == "0.1.0"


tests/test_pyanchorknit.py
from pyanchorknit import __version__


def test_version():
    assert __version__ == "0.3.0"


pyweaving/utils.py
"""Random utils."""
import logging

import typer


class TyperLoggerHandler(logging.Handler):
    def emit(self, record: logging.LogRecord) -> None:
        fg = None
        bg = None
        if record.levelno == logging.DEBUG:
            fg = typer.colors.BLACK
            bg = typer.colors.WHITE
        elif record.levelno == logging.INFO:
            fg = typer.colors.BRIGHT_BLUE
        elif record.levelno == logging.WARNING:
            fg = typer.colors.BRIGHT_MAGENTA
        elif record.levelno == logging.CRITICAL:
            fg = typer.colors.BRIGHT_RED
        elif record.levelno == logging.ERROR:
            fg = typer.colors.BRIGHT_WHITE
            bg = typer.colors.BRIGHT_RED
        typer.secho(self.format(record), bg=bg, fg=fg)


def config_logger(name: str, level: int = logging.DEBUG, to_file: bool = False):
    """Configure logging for stream and file."""
    logger = logging.getLogger(name)
    logger.setLevel(level)

    formatter = logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
    )

    # console_handler = logging.StreamHandler()
    # console_handler.setLevel(level)
    # console_handler.setFormatter(formatter)
    # logger.addHandler(console_handler)

    typer_handler = TyperLoggerHandler()
    typer_handler.setLevel(level)
    typer_handler.setFormatter(formatter)
    logger.addHandler(typer_handler)

    if to_file:
        try:
            file_handler = logging.FileHandler(f"{name}.log")
            file_handler.setLevel(logging.DEBUG)
            file_handler.setFormatter(formatter)
            logger.addHandler(file_handler)
        except OSError:
            pass

    return logger


pyanchorknit/cli/main.py
from pathlib import Path

import typer

from pyanchorknit.weaving import weaving

app = typer.Typer()


@app.command()
def main(
    img: Path = typer.Argument(..., help="Path to input image"),
    img_out: Path = typer.Option("./weaved.png", help="Path to output image"),
    traces_out: Path = typer.Option(
        "./traces.json", help="Path to output json of traces"
    ),
    n_edges: int = typer.Option(512, help="Number of points around the edge"),
    intensity_threshold: float = typer.Option(
        0.1,
        min=0,
        max=1,
        help="Min intensity threshold respect to the max intensity threshold to process",
    ),
    maxlines: int = typer.Option(2000, help="Max number of traces"),
    n_jobs: int = typer.Option(8, help="Max number of jobs to use"),
):
    """Make weaving image."""
    weaving(
        img,
        img_out=img_out,
        traces_out=traces_out,
        n_edges=n_edges,
        maxlines=maxlines,
        n_jobs=n_jobs,
        intensity_threshold=intensity_threshold,
    )


pyanchorknit/weaving.py
"""Weaving code."""
import json
from multiprocessing import Pool
from pathlib import Path
from typing import Dict, Tuple

import cairo
from cv2 import cv2
import numpy as np
from skimage.measure import profile_line
import tqdm

from pyanchorknit.utils import config_logger

logger = config_logger(__name__)

BLACK = 0
WHITE = 255


def whitest(intensities):
    return np.argmax(intensities)


def blackest(intensities):
    return np.argmin(intensities)


def create_circular_mask(h, w):
    center = (int(w / 2), int(h / 2))
    radius = min(center[0], center[1], w - center[0], h - center[1])

    Y, X = np.ogrid[:h, :w]
    dist_from_center = np.sqrt((X - center[0]) ** 2 + (Y - center[1]) ** 2)

    mask = dist_from_center <= radius
    return mask


def line_intensity(ps, points, bw_img):
    p1, p2 = ps
    return profile_line(bw_img.T, points[p1], points[p2], mode="constant").sum()


def weaving(
    img_path: Path,
    img_out: Path = Path("./weaved.png"),
    traces_out: Path = Path("./traces.json"),
    n_edges: int = 512,
    maxlines: int = 2000,
    n_jobs: int = 8,
    intensity_threshold: float = 0.1,
):
    logger.info(f"image path: {img_path}")
    logger.info(f"number of edges: {n_edges}")
    logger.info(f"max lines: {maxlines}")
    logger.info(f"n_jobs: {n_jobs}")

    img = cv2.imread(str(img_path.resolve()))
    bw_img = 255 - cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    centerx = img.shape[0] // 2
    centery = img.shape[1] // 2

    side = img.shape[0]
    halfside = side / 2

    points = []

    for theta in np.linspace(0, 2 * np.pi, n_edges + 1)[:-1]:
        x = min(int(centerx * np.cos(theta) + centerx), img.shape[0] - 1)
        y = min(int(centery * np.sin(theta) + centery), img.shape[1] - 1)
        points.append(np.array([x, y], dtype=np.int32))

    surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, *img.shape[:2])
    ctx = cairo.Context(surface)
    ctx.set_line_width(0.5)

    ctx.set_source_rgb(1, 1, 1)
    ctx.arc(halfside, halfside, halfside, 0, 2 * np.pi)
    ctx.clip()

    ctx.rectangle(0, 0, side, side)
    ctx.fill()

    ctx.set_source_rgb(0, 0, 0)

    traces: Dict[Tuple[int, int], float] = dict()
    thread_length = 0

    p = 0
    max_intensity = None

    ctx.move_to(*tuple(points[p]))

    for _ in tqdm.tqdm(range(maxlines)):
        plines = [(p, j) for j in range(len(points)) if j != p]

        with Pool(n_jobs) as pool:
            intensities = pool.starmap(
                line_intensity, [(ps, points, bw_img) for ps in plines]
            )

        poi = whitest(intensities)
        p, pprime = plines[poi]
        intensity = intensities[poi]

        max_intensity = max(max_intensity or 0, intensity)  # flip by color

        if len(traces) > 0 and (intensity / max_intensity) < intensity_threshold:
            logger.debug("Intensity below threshold")
            break

        if (p, pprime) in traces or (pprime, p) in traces:
            logger.debug("Repeated trace")
            break

        p1, p2 = tuple(points[p]), tuple(points[pprime])

        LINEWIDTH = 1
        bw_img = cv2.line(bw_img, p1, p2, BLACK, LINEWIDTH)

        ctx.line_to(*p2)

        thread_length += np.linalg.norm(points[p] - points[pprime])
        traces[(p, pprime)] = intensity
        p = pprime

    logger.info(f"Thread used: {thread_length:.2f}")

    ctx.stroke()
    # circular_mask = create_circular_mask(*img.shape[:2]) * 255
    # rgba_img = cv2.merge(np.array([white_img, white_img, white_img, circular_mask]))
    # cv2.imwrite(str(img_out.resolve()), rgba_img)
    with open(img_out, "wb") as fileobj:
        surface.write_to_png(fileobj)

    points_json = {i: tuple(map(int, point)) for i, point in enumerate(points)}
    traces_json = [((x, y), trace_len) for (x, y), trace_len in traces.items()]
    with open(traces_out, "w") as fp:
        json.dump({"points": points_json, "traces": traces_json}, fp)
